### DISCUSSION TITLE
Resource roadmap

### DISCUSSION THREAD

**Discussion by SkalskiP** [SkalskiP/courses#20] [Ideas]
I think one of the biggest problems with similar repositories like ours is:
- Too many resources that overwhelm people.
- Lack of direction. People don't know what they have to do first to achieve the goal they have set for themselves.

Therefore, I want to prepare a roadmap to show people which courses are most valuable. It will be a tree of knowledge - root: I know nothing, leaf: I have knowledge of the given topic. I need to refine my idea but it will be something like this:

```
├─ Foundation
│  ├─ 🧮 Math
│  │     ├─ course link
│  │     ├─ ...
│  │     └─ course link
│  └─ 👩🏻‍💻 Programming
│        ├─ course link
│        ├─ ...
│        └─ course link
.
.
.
```

**Comment by rasbt** [comment]
```python

To spin it further:

├─ Foundation
│  ├─ 🧮 Math
│  │     ├─ course link
│  │     ├─ ...
│  │     └─ course link
│  └─ 👩🏻‍💻 Programming
│        ├─ course link
│        ├─ ...
│        └─ course link
|
├─ AI Fundamentals
│  ├─  Machine Learning
│  │     ├─ course link
│  │     ├─ ...
│  │     └─ course link
│  └─ Deep Learning
│  │     ├─ course link
│  │     ├─  ...
│  │     ├─ course link
│  └─ AI
│        ├─ course link
│        ├─ ...
│        └─ course link
|
├─ Natural Language Processing
│       ├─ course link
│       ├─ ...
│       └─ course link
├─ Computer Vision
│       ├─ course link
│       ├─ ...
│       └─ course link
├─ Reinforcement Learning
│      ├─ course link
│      ├─ ...
│      └─ course link

.
.
.
```

Btw do you think generative AI should be a separate top-level category? It becomes tricky then because it intersects with many NLP and CV resources

**Reply by SkalskiP** [reply]
@rasbt I love it!

I think we can add generative AI as a separate category but put course links in multiple places. So, for example, the course on Stable Diffusion would be linked to `Generative AI` and `Computer Vision`—something like this.

```
│
├─ 🏗️ Foundation
│     ├─ 🧮 Math
│     │     ├─ course link
│     │     ├─ ...
│     │     └─ course link
│     └─ 👩🏻‍💻 Programming
│           ├─ course link
│           ├─ ...
│           └─ course link
│
├─ 🤖 AI Fundamentals
│     ├─  📈 Machine Learning
│     │     ├─ course link
│     │     ├─ ...
│     │     └─ course link
│     ├─ 🧠 Deep Learning
│     │     ├─ course link
│     │     ├─  ...
│     │     └─ course link
│     └─ 🧩 AI
│           ├─ course link
│           ├─ ...
│           └─ course link
│
├─ 💬 Natural Language Processing
│     ├─ course link
│     ├─ ...
│     └─ course link
│
├─ 📷 Computer Vision
│     ├─ [Stable Diffusion](https://course.fast.ai/Lessons/part2.html)
│     ├─ ...
│     └─ course link
│
├─ 🎮 Reinforcement Learning
│     ├─ course link
│     ├─ ...
│     └─ course link
│
├─ 🎨 Generative AI
│     ├─ [Stable Diffusion](https://course.fast.ai/Lessons/part2.html)
│     ├─ ...
│     └─ course link
.
.
.
```

Now the problem I see is that I can't have the monospace font and clickable hypertext links in markdown. Or can I?

**Reply by rasbt** [reply]
Oh yeah, that makes sense. I would perhaps put Generative AI above RL btw, because RL is really its own thing (and not really used for CV and NLP, except for PPO in some instruction-finetuned LLMs)

> Now the problem I see is that I can't have the monospace font and clickable hypertext links in markdown. Or can I?

Ah yeah, I don't think that's possible :(

**Reply by SkalskiP** [reply]
Haha, I think I found the solution. We will use HTML instead of the ``` section. Looks a bit stupid with those gigantic line breaks. But links work. 

<big><pre>
│
├─ 🏗️ Foundation
│     ├─ 🧮 Math
│     │     ├─ course link
│     │     ├─ ...
│     │     └─ course link
│     └─ 👩🏻‍💻 Programming
│           ├─ course link
│           ├─ ...
│           └─ course link
│
├─ 🤖 AI Fundamentals
│     ├─  📈 Machine Learning
│     │     ├─ course link
│     │     ├─ ...
│     │     └─ course link
│     ├─ 🧠 Deep Learning
│     │     ├─ course link
│     │     ├─  ...
│     │     └─ course link
│     └─ 🧩 AI
│           ├─ course link
│           ├─ ...
│           └─ course link
│
├─ 💬 Natural Language Processing
│     ├─ course link
│     ├─ ...
│     └─ course link
│
├─ 📷 Computer Vision
│     ├─ [Stable Diffusion](https://course.fast.ai/Lessons/part2.html)
│     ├─ ...
│     └─ course link
│
├─ 🎮 Reinforcement Learning
│     ├─ course link
│     ├─ ...
│     └─ course link
│
├─ 🎨 Generative AI
│     ├─ [Stable Diffusion](https://course.fast.ai/Lessons/part2.html)
│     ├─ ...
│     └─ course link
.
.
.
</pre></big>

**Reply by SkalskiP** [reply]
> Oh yeah, that makes sense. I would perhaps put Generative AI above RL btw, because RL is really its own thing (and not really used for CV and NLP, except for PPO in some instruction-finetuned LLMs)

Yeah, I'm not sure if `Natural Language Processing`, `Computer Vision`, `Reinforcement Learning` and `Generative AI` should all be on the same level. 

Curious if there is any other section we should add here. `Multimodal AI `?

**Reply by rasbt** [reply]
Oh yeah, the formatting doesn't look too bad!

Regarding the level structure ... Maybe NLP & CV should be under "Domain-Specific AI".  I would keep RL as a separate top-level one since it's really its own field. Maybe for Generative AI a top-level is ok too because then you can have NLP and CV also as sublevels there. 

Multimodal AI is tricky. I haven't watched the course and don't know what it covers. Is it strictly Generative AI-related? (Thinking of Stable Diffusion and other text2image models). 

Otherwise, maybe under "Special Topics", where Special Topics is a top-level category.

**Reply by SkalskiP** [reply]
> Oh yeah, the formatting doesn't look too bad!

I will dig deeper tomorrow. Maybe I'll find some better formatting options. I need to put it on pause. I need to do more work :) 

If we do not find better formatting options, I'll hide the learning resources roadmap in a clickable section like the one below so that it wouldn't hijack the whole `README.md`. I'll probably make some PR tomorrow.

<details close>
<summary>👉 learning resources roadmap</summary>

<big><pre>
│
├─ 🏗️ Foundation
│     ├─ 🧮 Math
│     │     ├─ course link
│     │     ├─ ...
│     │     └─ course link
│     └─ 👩🏻‍💻 Programming
│           ├─ course link
│           ├─ ...
│           └─ course link
│
├─ 🤖 AI Fundamentals
│     ├─  📈 Machine Learning
│     │     ├─ course link
│     │     ├─ ...
│     │     └─ course link
│     ├─ 🧠 Deep Learning
│     │     ├─ course link
│     │     ├─  ...
│     │     └─ course link
│     └─ 🧩 AI
│           ├─ course link
│           ├─ ...
│           └─ course link
│
├─ 💬 Natural Language Processing
│     ├─ course link
│     ├─ ...
│     └─ course link
│
├─ 📷 Computer Vision
│     ├─ [Stable Diffusion](https://course.fast.ai/Lessons/part2.html)
│     ├─ ...
│     └─ course link
│
├─ 🎮 Reinforcement Learning
│     ├─ course link
│     ├─ ...
│     └─ course link
│
├─ 🎨 Generative AI
│     ├─ [Stable Diffusion](https://course.fast.ai/Lessons/part2.html)
│     ├─ ...
│     └─ course link
.
.
.
</pre></big>

</details>

> Maybe NLP & CV should be under "Domain-Specific AI". I would keep RL as a separate top-level one since it's really its own field. Maybe for Generative AI a top-level is ok too because then you can have NLP and CV also as sublevels there.

I like that.

> Otherwise, maybe under "Special Topics", where Special Topics is a top-level category.

I like that too.

**Reply by SkalskiP** [reply]
The alternative is to drop monospace font and have this. But it looks a bit 💩 in my opinion.

- 🏗️ Foundation
    - 🧮 Math
        - [course link](https://example.com)
        - ...
        - [course link](https://example.com)
    - 👩🏻‍💻 Programming
        - [course link](https://example.com)
        - ...
        - [course link](https://example.com)
- 🤖 AI Fundamentals
    - 📈 Machine Learning
        - [course link](https://example.com)
        - ...
        - [course link](https://example.com)
    - 🧠 Deep Learning
        - [course link](https://example.com)
        - ...
        - [course link](https://example.com)
    - 🧩 AI
        - [course link](https://example.com)
        - ...
        - [course link](https://example.com)
- 💬 Natural Language Processing
    - [course link](https://example.com)
    - ...
    - [course link](https://example.com)
- 📷 Computer Vision
    - [course link](https://example.com)
    - ...
    - [course link](https://example.com)
- 🎮 Reinforcement Learning
    - [course link](https://example.com)
    - ...
    - [course link](https://example.com)
- 🎨 Generative AI
    - [course link](https://example.com)
    - ...
    - [course link](https://example.com)

**Reply by rasbt** [reply]
I actually don't mind this 😅

**Reply by SkalskiP** [reply]
Hah... maybe it is just me... If you could pick one:
- monospace font + links, but with bit line space
- font + links, but without monospace font 

Which would you choose?

**Comment by rasbt** [comment]
Don't want to make it even more complicated, but I also like the list style you currently have. So a 3rd option would be have is breaking the current list into smaller lists with different headers.

**Reply by SkalskiP** [reply]
Haha. This is cool but the problem I see is that when person is interested in CV they ask me what courses are prerequisite for course X. And I’d like to provide them not only with courses divided into groups but also the path they need to follow.
